File: 1621-number-of-sets-of-k-non-overlapping-line-segments/1621-number-of-sets-of-k-non-overlapping-line-segments.py
class Solution:
    def numberOfSets(self, n: int, k: int) -> int:
        MOD, prefix = 10 ** 9 + 7, [0] * (n + 1)
        dp = [1] * n

        for j in range(n):
            prefix[j + 1] = (prefix[j] + dp[j]) % MOD

        for _ in range(k):
            dp[0] = 0
            for j in range(1, n):
                dp[j] = (dp[j - 1] + prefix[j]) % MOD
            
            for j in range(n):
                prefix[j + 1] = (prefix[j] + dp[j]) % MOD

        return dp[n - 1]

File: 1621-number-of-sets-of-k-non-overlapping-line-segments/1621-number-of-sets-of-k-non-overlapping-line-segments.cpp
class Solution {
public:
    int numberOfSets(int n, int k) {
        vector<int> dp(n, 1), prefix(n + 1);
        int MOD = 1'000'000'007;

        for (int i = 0; i < n; ++i) prefix[i + 1] = (prefix[i] + dp[i]) % MOD;

        for (int i = 0; i < k; ++i) {
            dp[0] = 0;
            for (int j = 1; j < n; ++j) dp[j] = (dp[j - 1] + prefix[j]) % MOD;
            for (int j = 0; j < n; ++j) prefix[j + 1] = (prefix[j] + dp[j]) % MOD;
        }

        return dp.back();
    }
};